x-pack/plugins/cross_cluster_replication/server/routes/api/auto_follow_pattern.test.js

```diff
@@ -69,21 +72,39 @@ describe('[CCR API Routes] Auto Follow Pattern', () => {
       routeHandler = routeHandlers.list;
     });
 
-    it('should forward the response from Elasticsearch', async () => {
-      const mock = { foo: 'bar' };
-      setHttpRequestResponse(null, mock);
+    it('should deSerialize the response from Elasticsearch', async () => {
+      const totalResult = 2;
+      const deSerializedKeys = Object.keys(deserializeAutofollowPattern('random', getAutoFollowPatternMock()));
+      setHttpRequestResponse(null, getAutoFollowPatternListMock(totalResult));
 
       const response = await routeHandler();
+      const autoFollowPattern = Object.values(response)[0];
 
-      expect(response).toBe(mock);
+      expect(Object.keys(response).length).toEqual(totalResult);
+      expect(Object.keys(autoFollowPattern)).toEqual(deSerializedKeys);
     });
+  });
 
-    it('should prevent a 404 response on the route', async () => {
-      setHttpRequestResponse({ status: 404 });
-
-      const response = await routeHandler();
+  describe('create()', () => {
+    beforeEach(() => {
+      routeHandler = routeHandlers.create;
+    });
 
-      expect(response).toEqual({});
+    it('should serialize the payload before sending it to Elasticsearch', async () => {
+      callWithRequestFactory.mockReturnValueOnce((_, payload = {}) => {
+        if (payload.body.remote_cluster !== 'bar') {
+          return `Error: body (${JSON.stringify(payload)})`;
```

Comment:
We can leverage Jest's diffing feature by tweaking this to return the actual request itself and then making assertions against that:

```js
    it('should serialize the payload before sending it to Elasticsearch', async () => {
      // Return the payload we're sending to ES so we can make assertions against it.
      callWithRequestFactory.mockReturnValueOnce((_, payload = {}) => payload);

      const request = {
        params: { id: 'foo' },
        payload: {
          remoteCluster: 'bar1',
          leaderIndexPatterns: [ 'bar2' ],
          followIndexPattern: 'bar3',
        },
      };

      const response = await routeHandler(request);
      expect(response).toEqual({
        id: 'foo',
        body: {
          remote_cluster: 'bar1',
          leader_index_patterns: [ 'bar2' ],
          follow_index_pattern: 'bar3',
        },
      });
    });
```

This way I think the failures will be a little easier debug since they'll look something like this:

![image](https://user-images.githubusercontent.com/1238659/49046031-abccba00-f187-11e8-8de6-f6b03f023323.png)

Reply:
Good point, not sure why I went in that direction in the first place 😊